File: src/test/resources/oracle/commits/rm-action-test-288d1a960b5edef78540921f78809cafb3ef3ff6/typescript/OrderProcessor.ts
import { PricingPolicy } from "./PricingPolicy";
import { ReceiptFormatter } from "./ReceiptFormatter";

export class OrderProcessor {
    private readonly formatter = new ReceiptFormatter();

    buildReceipt(buyerName: string, unitPrice: number, quantity: number, discount: number): string {
        const total = this.calculateTotal(unitPrice, quantity, discount);
        const subtotal = unitPrice * quantity;
        const discountedSubtotal = subtotal - discount;
        const tax = Math.floor(discountedSubtotal * new PricingPolicy().taxRate);

        const header = this.formatter.createReceiptHeader(buyerName);
        const dollars = total / 100.0;
        const status = this.describePaymentStatus(total, discount);

        return (
            `${header}\n` +
            `Subtotal: ${subtotal}\n` +
            `Discount: ${discount}\n` +
            `Tax: ${tax}\n` +
            `Total: $${dollars}\n` +
            `${status}`
        );
    }

    private calculateTotal(unitPrice: number, quantity: number, discount: number): number {
        const subtotal = unitPrice * quantity;
        const discountedSubtotal = subtotal - discount;
        const tax = Math.floor(discountedSubtotal * new PricingPolicy().taxRate);
        return discountedSubtotal + tax;
    }

    private describePaymentStatus(total: number, discount: number): string {
        if (total > 0 && discount > 0) {
            return "Discount applied";
        }
        if (total > 0) {
            return "Payment required";
        }
        return "No payment required";
    }
}

File: src/test/resources/oracle/commits/rm-action-test-b65346600e43fad70b02f037dda796f86015f0c6/python/order_processor.py
from pricing_policy import PricingPolicy
from receipt_formatter import ReceiptFormatter


class OrderProcessor:
    def __init__(self):
        self._formatter = ReceiptFormatter()

    def build_receipt(self, buyer_name, unit_price, quantity, discount):
        total = self._calculate_total(unit_price, quantity, discount)
        subtotal = unit_price * quantity
        discounted_subtotal = subtotal - discount
        tax = int(discounted_subtotal * PricingPolicy().tax_rate)

        header = self._formatter.create_receipt_header(buyer_name)
        dollars = total / 100.0
        status = self._describe_payment_status(total, discount)

        return (
            f"{header}\n"
            f"Subtotal: {subtotal}\n"
            f"Discount: {discount}\n"
            f"Tax: {tax}\n"
            f"Total: ${dollars}\n"
            f"{status}"
        )

    def _calculate_total(self, unit_price, quantity, discount):
        subtotal = unit_price * quantity
        discounted_subtotal = subtotal - discount
        tax = int(discounted_subtotal * PricingPolicy().tax_rate)
        return discounted_subtotal + tax

    def _describe_payment_status(self, total, discount):
        if total > 0 and discount > 0:
            return "Discount applied"
        if total > 0:
            return "Payment required"
        return "No payment required"
